First conversation: https://chatgpt.com/share/66e3760f-2a60-8010-be5d-ea5aba055106

Second conversation: https://chatgpt.com/share/66e38a7d-6690-8010-8c8e-55fca2ea60ec

# Prelecture HW

#### 1. Importing dataset

In [26]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum() # counts NaN values 

The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.


row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

#### 2. Finding column and row counts

In [3]:
# Assuming your DataFrame is named df
rows, columns = df.shape

print(f'The DataFrame has {rows} rows and {columns} columns.')

The DataFrame has 391 rows and 11 columns.


Observations: number of individual entries in the dataset. In my case, this means number of rows, which means the dataset has 391 rows. 

Variables: attributes associated with each observation, which means all the column titles (row_n,id,name,gender,species,birthday,personality,song,phrase,full_id,url)

#### 3. Summarizing the columns 

In [6]:
df.head()

row_n       id     name  gender    species birthday personality  \
0      2  admiral  Admiral    male       bird     1-27      cranky   
1      3  agent-s  Agent S  female   squirrel      7-2       peppy   
2      4    agnes    Agnes  female        pig     4-21        uchi   
3      6       al       Al    male    gorilla    10-18        lazy   
4      7  alfonso  Alfonso    male  alligator      6-9        lazy   

          song    phrase           full_id  \
0   Steep Hill   aye aye  villager-admiral   
1      DJ K.K.  sidekick  villager-agent-s   
2   K.K. House   snuffle    villager-agnes   
3   Steep Hill   Ayyeeee       villager-al   
4  Forest Life  it'sa me  villager-alfonso   

                                                 url  
0  https://villagerdb.com/images/villagers/thumb/...  
1  https://villagerdb.com/images/villagers/thumb/...  
2  https://villagerdb.com/images/villagers/thumb/...  
3  https://villagerdb.com/images/villagers/thumb/...  
4  https://villagerdb.com/images/villagers/thumb/...

In [8]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

#### 4. Difference between `df.shape` and `df.describe`

Firstly, importing Titanic dataset as Animal Crossing dataset has no NaN numeric values. 

In [46]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.isna().sum() # counts NaN values 

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [20]:
df.shape

(891, 15)

In [22]:
df.describe(include='all')

survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male deck  embark_town alive  \
count   891.000000      889    891  891        891  203          889   891   
unique         NaN        3      3    3          2    7            3     2   
top            NaN        S  Third  man       True    C  Southampton    no   
freq           NaN      644    491  537        537   59          644   549   
mean     32.204208      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
std      49.693429      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
min       0.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
25%       7.910400      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
50%      14.454200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
75%      31.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
max     512.329200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   

       alone  
count    891  
unique     2  
top     True  
freq     537  
mean     NaN  
std      NaN  
min      NaN  
25%      NaN  
50%      NaN  
75%      NaN  
max      NaN

In [23]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

#### Conclusion: 

#### 4a) non-numeric variables
``df.shape`` returns the entire size/dimensions of the dataset, regardless of non-numeric variables. However, ``df.describe()`` skips over all columns with non-numeric variables, meaning it analyzes only the ones with numerical values in it. 

#### 4b) missing values in numeric variables
``df.shape`` returns the entire size/dimensions of the dataset, while the count described in ``df.describe()`` decreases in accordance to the number of NaN values in the column (e.g. age, which has count=714 due to 177 NaN values.  

#### 5. Difference between attribute and method 

The difference is that ``df.shape`` is a static, pre-existing property belonging to the df object, while ``df.describe()`` is a method that performs operations on the df object. 

# Here's a summary of our conversation:

1. **Pandas Functions**:
   - `df.isna().sum()` helps identify missing values in each column of a DataFrame by returning the count of `NaN` values.
   - To find the number of rows and columns in a DataFrame, you can use `df.shape`.

2. **Dataset Structure**:
   - Your dataset on Animal Crossing characters contains columns like `row_n`, `id`, `name`, `gender`, etc.
   - **Observations** refer to rows (each representing an individual character), and **variables** refer to columns (each representing an attribute of the characters).

3. **Summarizing a DataFrame**:
   - You can use `df.info()` for a basic overview, `df.describe()` for statistical summaries, `df.isna().sum()` for missing values, and `df.nunique()` for unique values in each column.

4. **New Dataset (Titanic)**:
   - In the Titanic dataset, `NaN` values can affect how `df.describe()` reports statistics compared to `df.shape`. While `df.shape` reports the total number of rows and columns, `df.describe()` shows only non-`NaN` counts for each column.

5. **Attributes vs. Methods**:
   - Attributes (e.g., `df.shape`) store information and are accessed directly, while methods (e.g., `df.describe()`) perform operations on the DataFrame and return results when called.

Let me know if you'd like further details on any of these points!

# Postlecture HW

#### 6. Definition of summary statistics

1. **'count'**: means the total number of non-NaN values in that observation

2. **'mean'**: calculates the average of the values in the observation

3. **'std'**: returns the standard deviation of the observation

4. **'min'**: returns the minimum value in the observation

5. **'25%'**: returns the value of the 25% percentile of the observation

6. **'50%'**: returns the value of the 50% percentile of the observation

7. **'75%'**: returns the value of the 75% percentile of the observation

8. **'max'**: returns the maximum value in the observation 

#### 7. ``df.dropna()`` vs ``del df['col']``

1. ``df.dropna()`` is preferred when a row needs to be dropped (by using ``df.dropna(axis=0)``) as ``del df['col']`` only works on columns. Also more preferred when there are a lot of columns/rows that need to be dealt with at the same time, which would make ``del df['col']`` slower as ``df.dropna()`` is automatic. 
2. ``del df['col']`` may be preferred when there is only one specific column that you want to get rid of, and you want to keep other columns with missing data.
3. By first applying ``del df['col']`` first, we can get rid of columns that are unwanted/are entirely missing. Otherwise, if the column is entirely missing, then using ``df.dropna()`` first will delete every row in the dataset.

#### **7.4:**

In [47]:
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [48]:
df.shape

(891, 15)

In [49]:
del df['deck'] 
del df['age']

In [50]:
df.head()

survived  pclass     sex  sibsp  parch     fare embarked  class    who  \
0         0       3    male      1      0   7.2500        S  Third    man   
1         1       1  female      1      0  71.2833        C  First  woman   
2         1       3  female      0      0   7.9250        S  Third  woman   
3         1       1  female      1      0  53.1000        S  First  woman   
4         0       3    male      0      0   8.0500        S  Third    man   

   adult_male  embark_town alive  alone  
0        True  Southampton    no  False  
1       False    Cherbourg   yes  False  
2       False  Southampton   yes   True  
3       False  Southampton   yes  False  
4        True  Southampton    no   True

In [51]:
df.shape

(891, 13)

In [54]:
df = df.dropna()
df.shape

(889, 13)

In [55]:
df.isna().sum()

survived       0
pclass         0
sex            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

#### Conclusion: 
Firstly, I used ``del df['col']`` on the "deck" and "age" columns, as ``df.isna().sum()`` returned that there were 177 missing values in "age" and 688 in "deck". This decreased the total columns of the dataset from 15 to 13. I think used ``df.dropna()`` to remove the remaining two rows with missing values in the "embarked" column. 

``df.shape`` beforehand: (891,15) 

``df.shape`` afterwards: (889,13)

# Here’s a summary of our entire conversation:

1. **Pandas Functions**:
   - **`df.isna().sum()`**: Counts missing values (`NaN`) in each column of a DataFrame.
   - **`df.shape`**: Returns the number of rows and columns in the DataFrame.

2. **Dataset Structure**:
   - **Observations**: Refers to rows in the DataFrame, each representing an individual record (e.g., an Animal Crossing character).
   - **Variables**: Refers to columns, each representing an attribute of the observations (e.g., `name`, `species`).

3. **Summarizing a DataFrame**:
   - **`df.info()`**: Provides a summary of the DataFrame including data types and non-null counts.
   - **`df.describe()`**: Returns descriptive statistics for numerical columns. With `include='all'`, it can provide statistics for all columns.
   - **`df.isna().sum()`**: Shows the count of missing values per column.
   - **`df.head()`** and **`df.tail()`**: Display the first and last few rows of the DataFrame.
   - **`df.nunique()`**: Counts unique values in each column.

4. **Handling Missing Values**:
   - **`df.dropna()`**: Removes rows or columns with missing values based on specified criteria.
   - **`del df['col']`**: Deletes a specific column from the DataFrame.

5. **Attributes vs. Methods**:
   - **Attributes** (e.g., `df.shape`): Provide static information and are accessed directly without parentheses.
   - **Methods** (e.g., `df.describe()`): Perform operations or computations and return results when called with parentheses.

6. **Order of Operations**:
   - Applying **`del df['col']`** before **`df.dropna()`** is important to remove irrelevant columns first, ensuring that `dropna()` only operates on relevant columns and simplifies the DataFrame.

7. **New Dataset (Titanic)**:
   - In the Titanic dataset, `NaN` values affect `df.describe()` by potentially lowering the count of non-missing values reported for each column compared to the total number of rows in `df.shape`.

This summary captures the key points from our conversation on pandas DataFrame operations, handling missing data, and understanding the distinction between attributes and methods.

#### **8.1**

In [58]:
df.groupby("pclass")["fare"].describe()
# where "pclass" = col1 and "fare" = col2

count       mean        std  min       25%      50%   75%       max
pclass                                                                     
1       214.0  84.193516  78.746457  0.0  30.77185  58.6896  93.5  512.3292
2       184.0  20.662183  13.417399  0.0  13.00000  14.2500  26.0   73.5000
3       491.0  13.675550  11.778142  0.0   7.75000   8.0500  15.5   69.5500

#### Explanation: 

``.describe()`` provides "count", "mean" and other summary statistics

``.groupby('col1')`` makes every distinct value in ``col1`` the first column/the independent variable, and ``['col2']`` makes ``col2`` the dependent variable that is being described by the summary statistics. 

#### **8.2**

In [59]:
# Reimporting the dataset 

import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.isna().sum() # counts NaN values 

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [60]:
df.groupby("pclass")['age'].describe()

count       mean        std   min   25%   50%   75%   max
pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

In [61]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

#### Explanation: 
``df.describe()`` returns summary statistics regarding the entire dataset, and that includes the value for "count". 

However, when used in conjunction with ``group_by``, it counts the number of non-NaN values within "col2", sorted in accordance to distinct "col1" values. 

#### **8.3:**

In [1]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

NameError: name 'pd' is not defined

A: Both were equally easy to fix. 

In [3]:
import pandas as pd 
url = "titanics.csv"
df = pd.read_csv(url)

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

B: ChatGPT was easier as it instantly gave me copy-pastable code to use, whereas Google led me to Stack Overflow which directed me to install an external library. 

In [4]:
import pandas as pd 
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

DF.describe()

NameError: name 'DF' is not defined

C: ChatGPT was easier as it directly pointed out that "DF" was the error; Google just said to check variable names which is much less precise 

In [5]:
df.describe(

SyntaxError: unexpected EOF while parsing (3431639942.py, line 1)

D: Again, ChatGPT easier as it directly pointed out that I was missing a parenthesis, whereas in Google I had to sort through the definition of the error and non-relevant examples others were posting. 

In [6]:
df.describle()

AttributeError: 'DataFrame' object has no attribute 'describle'

E: ChatGPT easier as it replaced "describle" with the precise function name; Google results were not relevant at all. 

In [8]:
df.groupby("sex")['Age'].describe()

KeyError: 'Column not found: Age'

In [9]:
df.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

F: Both were equally okay (however ChatGPT directed me to print out the columns, which was more helpful than Google merely telling me that I may have a typo)

In [10]:
df.groupby(sex)['age'].describe()

NameError: name 'sex' is not defined

G: ChatGPT much more helpful; immediately fixed the missing quotation marks whereas Google was not helpful at all, as NameError is too vague. 

# Here's a summary of your recent conversation:

1. **Initial Inquiry**: You were working on analyzing the Titanic dataset and asked for help understanding the code and analysis.
2. **Code Explanation**: I explained what `df.groupby("col1")["col2"].describe()` does, which is to provide descriptive statistics for `"col2"` grouped by `"col1"`.
3. **Error Handling**: You encountered a `NameError: name 'pd' is not defined` error and were advised to ensure you import pandas with `import pandas as pd`.
4. **FileNotFoundError**: You received a `FileNotFoundError` and were guided to check the file path and name, or use the correct URL for downloading data from the internet.
5. **Incorrect Method Name**: You had a `SyntaxError` and an `AttributeError` due to incorrect method names and were instructed to use `describe()` instead of `describle` and ensure correct syntax.
6. **Column Name Error**: You faced a `KeyError` for a column name and were advised to check the actual column names using `df.columns` and correct the column name in your `groupby` operation.
7. **NameError for Column Name**: You encountered a `NameError` when using a column name without quotes in `groupby`. It was clarified that column names need to be specified as strings in the `groupby` method.

If you have more questions or need further clarification, feel free to ask!